operators/src/opencl/cl_program.rs

```diff
@@ -404,31 +429,61 @@ impl<'a> CLProgramRunnable<'a> {
-                                std::slice::from_raw_parts(
-                                    coordinates.as_ptr() as *const Double2,
-                                    coordinates.len(),
-                                )
-                            })
-                            .build()?;
-
-                        kernel.set_arg(format!("IN_POINT_COORDS{}", idx), &buffer)?;
-
-                        let coordinates_offsets = points.multipoint_offsets();
-                        let buffer = Buffer::builder()
-                            .queue(queue.clone())
-                            .len(coordinates_offsets.len())
-                            .copy_host_slice(coordinates_offsets)
-                            .build()?;
-
-                        kernel.set_arg(format!("IN_POINT_OFFSETS{}", idx), &buffer)?;
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_COORDS", idx),
+                            &create_coordinate_buffer(queue, points.coordinates())?,
+                        )?;
+
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_OFFSETS", idx),
+                            &create_offset_buffer(queue, points.multipoint_offsets())?,
+                        )?;
+
+                        kernel.set_arg(format!("IN_FEATURE{}_LEN", idx), points.len() as i32)?;
+                    }
+                    TypedFeatureCollection::MultiLineString(multi_lines) => {
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_COORDS", idx),
+                            &create_coordinate_buffer(queue, multi_lines.coordinates())?,
+                        )?;
+
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_OFFSETS", idx),
+                            &create_offset_buffer(queue, multi_lines.multi_line_string_offsets())?,
+                        )?;
+                        kernel
+                            .set_arg(format!("IN_FEATURE{}_LEN", idx), multi_lines.len() as i32)?;
+
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_LINE_OFFSETS", idx),
+                            &create_offset_buffer(queue, multi_lines.line_string_offsets())?,
+                        )?;
+                    }
+                    TypedFeatureCollection::MultiPolygon(multi_polygons) => {
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_COORDS", idx),
+                            &create_coordinate_buffer(queue, multi_polygons.coordinates())?,
+                        )?;
+
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_OFFSETS", idx),
+                            &create_offset_buffer(queue, multi_polygons.multi_polygon_offsets())?,
+                        )?;
+                        kernel.set_arg(
+                            format!("IN_FEATURE{}_LEN", idx),
```

Comment:
don't you need offsets_len. ring_offsets_len and polygon_offsets_len and coords_len?

Reply:
One is enough to get the other length from within the kernel, cf. my test kernels.

Reply:
b9b625f